# Dual_autodiff

## Introduction

Consider the dual number *x = a + bε*. Then, *x² = (a + bε)² = a² + 2abε*.
Note that the derivative of *x²* with respect to *x* is *2ax*. Thus, with *b = 1*, the dual part of *x²*, i.e., *2abε*, is the derivative of *x²* with respect to *x* evaluated at *x = a*.  

In fact, you can show that any function *f(x)* can be extended to dual numbers and that its dual part is its derivative. You have:

*f(a + bε) = f(a) + f'(a)bε*

We refer to the part of the course on differentiable programming and the [Wikipedia page](https://en.wikipedia.org/wiki/Dual_number) for more details and references therein.

## User guide

### Installation

To install the package, install it using

In [2]:
import dual_autodiff

You can start using the module by:

In [3]:
from dual_autodiff import Dual

### Basic Usage

The package has one main module called Dual. In this module, many mathematical operations are ready to use. We also add some more complex functions such as hyperbolic functions. All you have to do is call the function, put in the correct inputs and it will run smooth as butter.

Here are some examples of how we make and define the function

## Examples

If you want to add two dual number, (6,2) and (3,1)

In [4]:
x = Dual(6,2)
y = Dual(3,1)

print(x+y)

Dual(real=9, dual=3)


Or if you want to see the multiplication of two dual number (6,2) and (3,1)

In [6]:
x = Dual(6,2)
y = Dual(3,1)

print(x*y)

Dual(real=18, dual=12)


Or the sin of the dual number (2,1) cannot be easier to find

In [5]:
x = Dual(2,1)
print(x.sin())

Dual(real=0.9092974268256817, dual=-0.4161468365471424)


In [11]:
import math 
z = math.sin(2)
z

0.9092974268256817

As you can see from here, if the dual component of a dual number x is 1 then the real component of the value of a function that is applied to x is actually the derivative of that function applied to the real part of x.

i.e, the real component of x.sin() is actually the derivative of sin() applied to the real component of x
i.e, sin(2) = 0.909

The package also allows you to conduct exponential function

In [12]:
x = Dual(2,1)
print(x.exp())

Dual(real=7.38905609893065, dual=7.38905609893065)


and also logarithm if you want it the other way round

In [13]:
x = Dual(7.389, 7.389)
print(x.log())

Dual(real=1.9999924078065106, dual=1.0)


### Errors

To use the package or make it run properly, inputs should be handled with care.

For example, normal mathematical properties are applied. We cannot take the logarithm of a non-positive real part:

In [14]:
x = Dual(-1, 3)
print(x.log())

ValueError: Log undefined for non-positive.

And we cannot calculate the tan value of a dual number that has cos() is equal to 0

In [16]:
x = Dual(math.pi/2, 1)
print(x.tan())

ValueError: Tangent is undefined for cos(real) = 0.

Or we don't want you to put in your real part only:

In [17]:
x = Dual(1)
print(x)

TypeError: __init__() missing 1 required positional argument: 'dual'

And of course we cannot find any dual numbers with letters/string real or dual component

In [19]:
x = Dual(1, "one")
print(x)

TypeError: Dual part must be a numeric type, got str

There are more to explore using the package, I hope you have fun playing with it!